frontend/test/metabase/lib/dataset.unit.spec.js

```diff
@@ -93,6 +94,80 @@ describe("metabase/util/dataset", () => {
       ]);
       expect(newQuestion.query().columns()).toHaveLength(4);
     });
+
+    it("removes columns from table.columns when a column is removed from a query", () => {
+      const prevQuestion = productsTable
+        .query({
+          fields: [
+            ["field", PRODUCTS.ID, null],
+            ["field", PRODUCTS.CATEGORY, null],
+            ["field", PRODUCTS.VENDOR, null],
+          ],
+        })
+        .question()
+        .setSettings({
+          "table.columns": [
+            createMockTableColumnOrderSetting({
+              name: "ID",
+              fieldRef: ["field", PRODUCTS.ID, null],
+              enabled: true,
+            }),
+            createMockTableColumnOrderSetting({
+              name: "CATEGORY",
+              fieldRef: ["field", PRODUCTS.CATEGORY, null],
+              enabled: true,
+            }),
+            createMockTableColumnOrderSetting({
+              name: "VENDOR",
+              fieldRef: ["field", PRODUCTS.VENDOR, null],
+              enabled: true,
+            }),
+          ],
+        });
+
+      const newQuestion = prevQuestion
+        .query()
+        .removeField(2)
+        .question()
+        .syncColumnsAndSettings(prevQuestion);
+
+      expect(prevQuestion.setting("table.columns")).toHaveLength(3);
+      expect(newQuestion.setting("table.columns")).toHaveLength(2);
```

Comment:
Could we assert which `table.columns` are left in `newQuestion`? Things would be more reliable in that case — e.g., the test would still pass if we remove a wrong column